### PR TITLE
Sagepay one-off payments

### PR DESCRIPTION
**Pull request by elcapo**
Hello @eileenmcnaughton,

Here is my first attempt to make Sagepay payments work. It only enables one-off payments (I'll be working on recurring ones but they're not covered yet).

Can you let me know if these changes look correct to you?

## Notification from different session

I added a `notification_from_different_session` payment processor metadata to identify a Sagepay behavior: contributions are completed by their notification services and they redirect the user to CiviCRM after the payment is completed.

As the qfKey is stored in the user's browser session and Sagepay also requires us remembering a `SecurityKey`, both the qfKey and the SecurityKey are (pre)-stored in the contribution and then queried from there during the notification process.

## Shipping fields can't be empty

To prevent a `3140 : The DeliveryCountry value is invalid`, I copied them from the billing data, just in case they are missing.

### CODE REVIEW

File: CRM/Core/Payment/OmnipayMultiProcessor.php [hunk 1]
```diff
@@ -512,6 +577,8 @@ private function getCreditCardObjectParams($params) {
       $cardFields['billingState'] = CRM_Core_PseudoConstant::stateProvinceAbbreviation($cardFields['billingState']);
     }
 
+    $this->copyShippingFieldsFromBillingIfEmpty($cardFields);
```
Comment (mattwire) [inline CRM/Core/Payment/OmnipayMultiProcessor.php]: @elcapo At least with sagepay there are two ways to handle the missing shipping fields. Either copy them or set `billingForShipping` to 1 in the omnipay sagepay gateway params - see https://github.com/eileenmcnaughton/nz.co.fuzion.omnipaymultiprocessor/pull/42/files#diff-71ff8f521b77ef1a6c6643105f8b24d5R60

Comment (elcapo) [reply CRM/Core/Payment/OmnipayMultiProcessor.php]: I've just tried removing the `copyShippingFieldsFromBillingIfEmpty` function in favor of the `gateway_params` you suggest but that takes me again to the _3140 : The DeliveryCountry value is invalid_ error message. Also, I didn't find any usage of the `gateway_params` in the extension. Maybe it has been deprecated?

Comment (mattwire) [reply CRM/Core/Payment/OmnipayMultiProcessor.php]: It wasn't fully implemented in this extension yet - see https://github.com/eileenmcnaughton/nz.co.fuzion.omnipaymultiprocessor/pull/81

**Comment by elcapo** [comment]
> Also, @elcapo can you confirm if this PR can be closed or would still be needed with your changes? #82

It can be closed from my point of view. Only thing I see I'm not doing yet is in this PR is to complete an upgrade to Sagepay 3 by removing the `send` part from:

```
$this->gateway->acceptNotification($params)->send();
```

But I guess that can be done in a different PR (fully compliance with Omnipay 3.0 or something like that).

By the way, I'll be sending (hopefully this week) a PR with support for repeated transactions using continuous authority support.

**Comment by elcapo** [comment]
@mattwire Just in case you want to start testing repeated payments through continuous authority, I've a "working draft" in this [continuous_authority](https://github.com/elcapo/nz.co.fuzion.omnipaymultiprocessor/tree/continuous_authority) branch. I'll be working on it to have ready a PR soon. Any feedback is welcome.

File: CRM/Core/Payment/OmnipayMultiProcessor.php (hunk 1 above)
Comment (elcapo) [reply CRM/Core/Payment/OmnipayMultiProcessor.php]: Thanks! Good to know! I'll do that your way when your PR is accepted. It looks cleaner than mine.

**Comment by elcapo** [comment]
@eileenmcnaughton If you see this, please consider accepting @mattwire 's [Pull Request about gateway params via metadata](https://github.com/eileenmcnaughton/nz.co.fuzion.omnipaymultiprocessor/pull/81) first. I've a update ready for this Pull Request where, assuming Matt's PR is there, I can remove the `copyShippingFieldsFromBillingIfEmpty` function, which was actually breaking some tests.

**Comment by mattwire** [comment]
@elcapo This needs a rebase - maybe best to include #81 in this PR?

**Comment by eileenmcnaughton** [comment]
I'm closing this out now as the functionality for it has been merged in as well as the functionality to make recurring payments work - it seems to work fine in my tests but needs more testing before I tag a release

**Pull request closed without merging**